### ISSUE TITLE
TestGOAWAYConcurrency fails with data race issues

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#94532]
This fails close to 100% of the time when run with race mode:

```
go test -race k8s.io/apiserver/pkg/server/filters -run TestGOAWAYConcurrency -v
```

Disabled in https://github.com/kubernetes/kubernetes/pull/94529

It is unclear if the test or the component needs to be corrected

/sig api-machinery
/priority critical-urgent

**Comment by knight42** [comment]
@answer1991 The test `TestClientReceivedGOAWAY/combine_with_watch-with-goaway` is also flaky.

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/periodic-kubernetes-bazel-test-master/1302013439975100416

```
=== RUN   TestClientReceivedGOAWAY
=== RUN   TestClientReceivedGOAWAY/all_normal_requests_use_only_one_connection
=== RUN   TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch
=== RUN   TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch,_and_set-up_a_new_watch
=== RUN   TestClientReceivedGOAWAY/got_2_GOAWAY_after_set-up_watch
=== RUN   TestClientReceivedGOAWAY/combine_with_watch-with-goaway
    goaway_test.go:377: expect TCP connection: 3, actual: 4
--- FAIL: TestClientReceivedGOAWAY (20.11s)
```

but this one is hard to reproduce locally.

**Comment by answer1991** [comment]
I will debug and try to fix it later, thanks for the report. @knight42

**Comment by answer1991** [comment]
I doubt `TestGOAWAYConcurrency` data race is caused by golang http/http2 pkg, let me try to make some reproduce cases and may open an issue to golang repo.

**Comment by answer1991** [comment]
`TestGOAWAYConcurrency` data race is caused by golang http/http2 pkg, I have already opened an issue [#41234](https://github.com/golang/go/issues/41234) at golang repo.

I will continue to debug `TestClientReceivedGOAWAY/combine_with_watch-with-goaway` tomorrow.

**Comment by knight42** [comment]
another case: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/94556/pull-kubernetes-bazel-test/1302300454738202625
```
=== RUN   TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch,_and_set-up_a_new_watch
    goaway_test.go:377: expect TCP connection: 2, actual: 3
=== RUN   TestClientReceivedGOAWAY/got_2_GOAWAY_after_set-up_watch
=== RUN   TestClientReceivedGOAWAY/combine_with_watch-with-goaway
--- FAIL: TestClientReceivedGOAWAY (15.10s)
    --- PASS: TestClientReceivedGOAWAY/all_normal_requests_use_only_one_connection (0.01s)
    --- PASS: TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch (5.01s)
    --- FAIL: TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch,_and_set-up_a_new_watch (0.03s)
    --- PASS: TestClientReceivedGOAWAY/got_2_GOAWAY_after_set-up_watch (5.01s)
    --- PASS: TestClientReceivedGOAWAY/combine_with_watch-with-goaway (5.02s)
```

**Comment by answer1991** [comment]
For `TestClientReceivedGOAWAY` case, change `tc.expectConnections != len(localAddr)` to be `tc.expectConnections > len(localAddr)` would be a solution:

https://github.com/kubernetes/kubernetes/blob/74053d555d71a14e3853b97e204d7d6415521375/staging/src/k8s.io/apiserver/pkg/server/filters/goaway_test.go#L376-L379

But I would rather figure out why client will set-up the additional TCP connections.

**Comment by answer1991** [comment]
Fix TestGOAWAYConcurrency flaky may need a new golang version after golang/go#41234 fixed.

TestClientReceivedGOAWAY will be tracked at #94622 .

**Comment by lavalamp** [comment]
Great find-- can we skip the test for now while we wait for the go fix to land?

**Comment by answer1991** [comment]
> can we skip the test for now while we wait for the go fix to land?

Yes, I think so.

**Comment by answer1991** [comment]
/lifecycle stale

I'm still working on it, a new golang version(1.16 or 1.15.7+) release required

**Comment by answer1991** [comment]
> /lifecycle stale
> 
> I'm still working on it, a new golang version(1.16 or 1.15.7+) release required

As clients in k/k repo use [golang.org/x/net/http2](https://github.com/kubernetes/kubernetes/blob/c9604bd302f9ff4df6c4cf6a8cc9130a497bcf36/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L174) to config http2 transports instead of use golang SDK h2_bundle, a new version of golang.org/x/net(with https://github.com/golang/net/commit/ff519b6c91021e6316e1df005bc19f266994ddda) will fix this issue.

However, for other clients which use default http2 configuration in golang SDK, golang 1.15.8 or 1.16+ required to avoid HTTP2 un-expected error when received GOAWAY frame.

**Comment by BenTheElder** [comment]
1.15.8 should be landed now https://github.com/kubernetes/kubernetes/pull/98834

**Comment by answer1991** [comment]
yes, I have already post a pr to enable goaway feature test cases: https://github.com/kubernetes/kubernetes/pull/98809
